examples/pipeline/wav2letter.py

```diff
@@ -0,0 +1,1179 @@
+    "eps": args.eps,
+    "rho": args.rho,
+    "weight_decay": args.weight_decay,
+}
+
+optimizer_params_adam = {
+    "lr": args.learning_rate,
+    "eps": args.eps,
+    "weight_decay": args.weight_decay,
+}
+
+optimizer_params_sgd = {
+    "lr": args.learning_rate,
+    "weight_decay": args.weight_decay,
+}
+
+optimizer_params_adadelta = {
+    "lr": args.learning_rate,
+    "eps": args.eps,
+    "rho": args.rho,
+    "weight_decay": args.weight_decay,
+}
+
+Optimizer = Adadelta
+optimizer_params = optimizer_params_sgd
+
+# Model
+
+num_features = n_bins if n_bins else 1
+
+lstm_params = {
+    "hidden_size": 800,
+    "num_layers": 5,
+    "batch_first": False,
+    "bidirectional": False,
+    "dropout": 0.,
+}
+
+clip_norm = 0.  # 10.
+
+zero_infinity = False
+
+
+# Text encoding
+
+class Coder:
+    def __init__(self, labels):
+        labels = [l for l in labels]
+        self.length = len(labels)
+        enumerated = list(enumerate(labels))
+        flipped = [(sub[1], sub[0]) for sub in enumerated]
+
+        d1 = collections.OrderedDict(enumerated)
+        d2 = collections.OrderedDict(flipped)
+        self.mapping = {**d1, **d2}
+
+    def encode(self, iterable):
+        if isinstance(iterable, list):
+            return [self.encode(i) for i in iterable]
+        else:
+            return [self.mapping[i] + self.mapping[char_blank] for i in iterable]
+
+    def decode(self, tensor):
+        if isinstance(tensor[0], list):
+            return [self.decode(t) for t in tensor]
+        else:
+            # not idempotent, since clean string
+            x = (self.mapping[i] for i in tensor)
+            x = ''.join(i for i, _ in itertools.groupby(x))
+            x = x.replace(char_blank, "")
+            # x = x.strip()
+            return x
+
+
+coder = Coder(labels)
+encode = coder.encode
+decode = coder.decode
+vocab_size = coder.length
+print("vocab_size", vocab_size, flush=True)
+
+
+# Model
+
+model = Wav2Letter(num_features, vocab_size)
+
+def model_length_function(tensor):
+    return int(tensor.shape[0])//2 + 1
+
+
+# Dataset
+
+
+class IterableMemoryCache:
+
+    def __init__(self, iterable):
+        self.iterable = iterable
+        self._iter = iter(iterable)
+        self._done = False
+        self._values = []
+
+    def __iter__(self):
+        if self._done:
+            return iter(self._values)
+        return itertools.chain(self._values, self._gen_iter())
+
+    def _gen_iter(self):
+        for new_value in self._iter:
+            self._values.append(new_value)
+            yield new_value
+        self._done = True
+
+    def __len__(self):
+        return len(self._iterable)
+
+
+class MapMemoryCache(torch.utils.data.Dataset):
+    """
+    Wrap a dataset so that, whenever a new item is returned, it is saved to memory.
+    """
+
+    def __init__(self, dataset):
+        self.dataset = dataset
+        self._cache = [None] * len(dataset)
+
+    def __getitem__(self, n):
+        if self._cache[n]:
```

Comment:
To be pedantic, you could have dataset setup to return "False" (which I know it isn't in this case) and then this check would fail. To be more exact you'd use ```if self._cache[n] is not None:```.